### ISSUE TITLE
Proxy socket.io

### ISSUE THREAD

**Issue by greguz** [fastify/fastify-http-proxy#121]
## 🚀 Feature Proposal

I came across the necessity to proxy a Socket.io connection. Knowing that the socket support is partial, I tried to use this module as-is, but without success.

## Motivation

Not sure if Socket.io support can be a goal (because of the almost universal support of raw WebSockets in any environment).

## Example

Firstly I tried to emulate the module code: declaring my WebSocket instance and using the stream created by ws. This code does not throw, but the first received message is lost (saw while debugging the Socket.io server).

```javascript
import make from 'fastify-plugin'
import websocket from 'fastify-websocket'
import WebSocket from 'ws'
import { pipeline } from 'stream'

async function plugin(fastify) {
  fastify.register(websocket)

  fastify.route({
    method: 'GET',
    url: '/socket.io',
    handler(request, reply) {
      reply.code(404).send()
    },
    wsHandler(connection, request) {
      const ws = new WebSocket(process.env.SERVER + request.url, {
        perMessageDeflate: {
          threshold: 1024,
        },
        rejectUnauthorized: false,
      })

      const stream = WebSocket.createWebSocketStream(ws)

      fastify.log.info('starting websocket tunnel')
      pipeline(connection, stream, connection, function (err) {
        if (err) {
          fastify.log.info({ err }, 'websocket tunnel terminated with error')
          return
        }
        fastify.log.info('websocket tunnel terminated')
      })
    },
  })
}

export default make(plugin)
```
Instead, by proxying all events socket-to-socket, everything works.

```javascript
import make from 'fastify-plugin'
import websocket from 'fastify-websocket'
import WebSocket from 'ws'

function closeWebSocket(socket, code, reason) {
  if (socket.readyState === WebSocket.OPEN) {
    socket.close(code, reason)
  }
}

function proxy(a, b) {
  function close(code, reason) {
    closeWebSocket(a, code, reason)
    closeWebSocket(b, code, reason)
  }

  a.on('message', data => b.send(data))
  a.on('ping', data => b.ping(data))
  a.on('pong', data => b.pong(data))
  // cannot proxy error code because of ws
  a.on('close', (code, reason) => close(4600, reason))
  a.on('error', error => close(4610, error.message))
  a.on('unexpected-response', (req, res) =>
    close(4620, 'websocket upgrade failed'),
  )

  b.on('message', data => a.send(data))
  b.on('ping', data => a.ping(data))
  b.on('pong', data => a.pong(data))
  // cannot proxy error code because of ws
  b.on('close', (code, reason) => close(4600, reason))
  b.on('error', error => close(4610, error.message))
  b.on('unexpected-response', (req, res) =>
    close(4620, 'websocket upgrade failed'),
  )
}

async function plugin(fastify) {
  fastify.register(websocket)

  fastify.route({
    method: 'GET',
    url: '/socket.io',
    handler(request, reply) {
      reply.code(404).send()
    },
    wsHandler(connection, request) {
      proxy(
        connection.socket,
        new WebSocket(process.env.SERVER + request.url, {
          perMessageDeflate: {
            threshold: 1024,
          },
          rejectUnauthorized: false,
        }),
      )
    },
  })
}

export default make(plugin)
```

**Comment by fox1t** [comment]
@mcollina do you have any idea why we need to explicitly call the events between the two sockets? Is there any documentation on how this should work in Node.js? It is pretty surprising  for me that https://github.com/fastify/fastify-http-proxy/blob/master/index.js#L96 isn't enough.

**Comment by mcollina** [comment]
This makes sense. WebSockets have both data and metadata on the data framing. The Stream _decodes_ the framing, and then we _encode_ it again. It would be amazing to see if we could avoid this entirely, i.e. skip fastify-websocket.

**Comment by greguz** [comment]
@mcollina ok, I'll make a merge request in the next days.

**Comment by greguz** [comment]
Created MR #122. I tried to add the "rewrite headers" feature, but I found some problems with `Sec-WebSocket-*` headers. I suppose that those headers (and maybe others too) need to be removed while proxying the request.

**Comment by mekwall** [comment]
Was this fixed by #122?

**Comment by mcollina** [comment]
Yes!
